# Test Notebook For PyVO Client Testing with synchronous calls to KOA TAP Server v1.0  #

## Created by Bruce Berriman, April 7 2020 #

### Set up
#### Install PyVO via  $pip install PyVO 
#### The SQL embedded in the URLs has been decoded with http://www.asiteaboutnothing.net/c_decode-url.html 



In [1]:
from pyvo.dal import tap 

In [2]:
koa = tap.TAPService('http://vmkoatest.ipac.caltech.edu/TAP')

In [3]:
# import time

### Test 1- output_format_1


In [63]:
sql = "select koaid, filehand from koa_hires where koaid like '%20040824%' "


In [64]:
results_output_format_1= koa.run_sync(sql)
print(results_output_format_1)

<Table length=105>
        koaid                                filehand                      
        object                                object                       
---------------------- ----------------------------------------------------
HI.20040824.08204.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08204.fits
HI.20040824.08544.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08544.fits
HI.20040824.08591.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08591.fits
HI.20040824.08638.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08638.fits
HI.20040824.08685.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08685.fits
HI.20040824.08732.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08732.fits
HI.20040824.08779.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08779.fits
HI.20040824.08826.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08826.fits
HI.20040824.08873.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08873.fits
HI.20040824.08919.fits /koadata1/HIRES/20040824/lev0/HI.20040824.0891

## Test  2 - different output formats
#### PyVO may not be able accept formats other than voformat, so we will test that we can convert the VOTable to other formats.



In [68]:
##repeat frst test

sql = "select koaid, filehand from koa_hires where koaid like '%20040824%' "

In [69]:
##results
results = koa.run_sync(sql)
print, results

(<function print>, <Table length=105>
         koaid                                filehand                      
         object                                object                       
 ---------------------- ----------------------------------------------------
 HI.20040824.08204.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08204.fits
 HI.20040824.08544.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08544.fits
 HI.20040824.08591.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08591.fits
 HI.20040824.08638.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08638.fits
 HI.20040824.08685.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08685.fits
 HI.20040824.08732.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08732.fits
 HI.20040824.08779.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08779.fits
 HI.20040824.08826.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08826.fits
 HI.20040824.08873.fits /koadata1/HIRES/20040824/lev0/HI.20040824.08873.fits
 HI.20040824.08919.fits /koadata1/HIRE

In [70]:
# write results to an ascii table
table=results.to_table()
table.write ('./table_ipacascii.tbl',format='ascii.ipac',overwrite=True)

In [71]:
#  write results to CSV file
results = koa.run_sync(sql)
table=results.to_table()
table.write ('./table_csv.tbl',format='csv', overwrite=True)

### PASS

### Test  3:  Select all keywords


In [72]:
sql = "select * from koa_hires where koaid like '%20040824%' "

In [73]:
results = koa.run_sync(sql)
print, results

(<function print>, <Table length=105>
    el   elaptime equinox eramode ...      z        spt_ind      ofname    
 float64  int32   float64  object ...   float64      int32       object    
 ------- -------- ------- ------- ... ------------ --------- --------------
    45.0        1     0.0 B, G, R ... 0.7068576891 203233131 hires2300.fits
    45.0        1     0.0 B, G, R ... 0.7068576891 203233131 hires2301.fits
    45.0        1     0.0 B, G, R ... 0.7068576891 203233131 hires2302.fits
    45.0        1     0.0 B, G, R ... 0.7068576891 203233131 hires2303.fits
    45.0        1     0.0 B, G, R ... 0.7068576891 203233131 hires2304.fits
    45.0        1     0.0 B, G, R ... 0.7068576891 203233131 hires2305.fits
    45.0        1     0.0 B, G, R ... 0.7068576891 203233131 hires2306.fits
    45.0        1     0.0 B, G, R ... 0.7068576891 203233131 hires2307.fits
    45.0        1     0.0 B, G, R ... 0.7068576891 203233131 hires2308.fits
    45.0        1     0.0 B, G, R ... 0.7068576891

In [74]:
## write output file
table_selectall=results.to_table()
table_selectall.write ('./table_selectall.tbl',format='votable',overwrite=True)

### Pass

### Test 4: Select_2  - select filehandle keyword


In [75]:
sql = "select filehand from koa_hires where koaid like '%20040824%' "
results = koa.run_sync(sql)
print, results

(<function print>, <Table length=105>
                       filehand                      
                        object                       
 ----------------------------------------------------
 /koadata1/HIRES/20040824/lev0/HI.20040824.08204.fits
 /koadata1/HIRES/20040824/lev0/HI.20040824.08544.fits
 /koadata1/HIRES/20040824/lev0/HI.20040824.08591.fits
 /koadata1/HIRES/20040824/lev0/HI.20040824.08638.fits
 /koadata1/HIRES/20040824/lev0/HI.20040824.08685.fits
 /koadata1/HIRES/20040824/lev0/HI.20040824.08732.fits
 /koadata1/HIRES/20040824/lev0/HI.20040824.08779.fits
 /koadata1/HIRES/20040824/lev0/HI.20040824.08826.fits
 /koadata1/HIRES/20040824/lev0/HI.20040824.08873.fits
 /koadata1/HIRES/20040824/lev0/HI.20040824.08919.fits
                                                  ...
 /koadata1/HIRES/20040824/lev0/HI.20040824.54803.fits
 /koadata1/HIRES/20040824/lev0/HI.20040824.54849.fits
 /koadata1/HIRES/20040824/lev0/HI.20040824.54896.fits
 /koadata1/HIRES/20040824/lev0/HI.20040824.5

In [76]:
# write output file
table_selectfilehand=results.to_table()
table.write ('./table_selectfilehand.tbl',format='votable',overwrite='True')

### PASS

### Test 5: Select keyword that does not exist
#### select%20DOESNOTEXIST%20from%20koa.koa_hires_v7%20where%20koaid%20like%20%27%2520040824%25%27&format=ipac

In [77]:
sql = "select DOESNOTEXIST from koa_hires where koaid like '%20040824%' "

try:
    results = koa.run_sync(sql)
    table   = results.to_table()
    table.write('./select_DOES.tbl',format='ascii.ipac',overwrite=True)
    
except Exception as e:
    with open ('./select_DOES.tbl', 'w+') as fp:
      fp.write (f'Error: {str(e):s}')
      fp.flush()
    

#### PASS Returns correct error message

### Test 6: select KOAID twice


In [78]:
sql = "select koaid, koaid from koa_hires where koaid like '%20040824%' "
results = koa.run_sync(sql)

In [80]:
print (results)


<Table length=105>
        koaid                  koaid2        
        object                 object        
---------------------- ----------------------
HI.20040824.08204.fits HI.20040824.08204.fits
HI.20040824.08544.fits HI.20040824.08544.fits
HI.20040824.08591.fits HI.20040824.08591.fits
HI.20040824.08638.fits HI.20040824.08638.fits
HI.20040824.08685.fits HI.20040824.08685.fits
HI.20040824.08732.fits HI.20040824.08732.fits
HI.20040824.08779.fits HI.20040824.08779.fits
HI.20040824.08826.fits HI.20040824.08826.fits
HI.20040824.08873.fits HI.20040824.08873.fits
HI.20040824.08919.fits HI.20040824.08919.fits
                   ...                    ...
HI.20040824.54803.fits HI.20040824.54803.fits
HI.20040824.54849.fits HI.20040824.54849.fits
HI.20040824.54896.fits HI.20040824.54896.fits
HI.20040824.54944.fits HI.20040824.54944.fits
HI.20040824.54990.fits HI.20040824.54990.fits
HI.20040824.55037.fits HI.20040824.55037.fits
HI.20040824.55084.fits HI.20040824.55084.fits
HI.20040824.551

In [20]:
## write to output file
table_2koaid=results.to_table()
table.write ('./table_2koaid.tbl',format='votable',overwrite='True')

#### PASS - but renames 2nd KOAID

### Test 6: select order by sig2nois


In [81]:
sql = "select koaid, koaid from koa_hires where koaid like '%20040824%' order by sig2nois "
results = koa.run_sync(sql)

In [82]:
## write output file
table_test6_orderbysig2nois=results.to_table()
table.write ('./table_test6_orderbysig2nois.tbl',format='votable',overwrite='True')

In [83]:
sql = "select koaid, filehand, sig2nois from koa_hires where koaid like '%20040824%'order by sig2nois  "
results = koa.run_sync(sql)
print,results

(<function print>, <Table length=105>
         koaid          ... sig2nois
         object         ... float64 
 ---------------------- ... --------
 HI.20040824.09677.fits ...      2.0
 HI.20040824.09724.fits ...      2.0
 HI.20040824.54001.fits ...      2.0
 HI.20040824.53947.fits ...      2.0
 HI.20040824.45942.fits ...      2.0
 HI.20040824.44695.fits ...      2.0
 HI.20040824.44395.fits ...      2.0
 HI.20040824.44343.fits ...      2.0
 HI.20040824.43054.fits ...      2.0
 HI.20040824.41806.fits ...      2.0
                    ... ...      ...
 HI.20040824.09011.fits ...     11.0
 HI.20040824.08965.fits ...     11.0
 HI.20040824.08873.fits ...     11.0
 HI.20040824.09105.fits ...     11.0
 HI.20040824.08779.fits ...     11.0
 HI.20040824.08732.fits ...     11.0
 HI.20040824.08826.fits ...     11.0
 HI.20040824.08591.fits ...     11.0
 HI.20040824.08204.fits ...     11.0
 HI.20040824.08685.fits ...     15.0)

### PASS

### Test 7: order by DOESNOTEXIST
#### select%20koaid,%20filehand,%20sig2nois%20from%20koa.koa_hires_v7%20where%20koaid%20like%20%27%2520040824%25%27%20order%20by%20DOESNOTEXIST&format=ipac

In [84]:
sql = "select koaid, filehand, sig2nois from koa_hires where koaid like '%20040824%'order by DOESNOTEXIST  "

try:
    results = koa.run_sync(sql)
    table   = results.to_table()
    table.write('./select_orderdoesnot.tbl',format='ascii.ipac',overwrite=True)
    
except Exception as e:
    with open ('./select_orderbydoesnot.tbl', 'w+') as fp:
      fp.write (f'Error: {str(e):s}')
      fp.flush()

### PASS returns error condtion

### Test 8: set max records =3



In [85]:
sql="select koaid, filehand, frameno from koa_hires where koaid like '%20040824%'"
results = koa.run_sync(sql, maxrec=3)
print, results

(<function print>, <Table length=3>
         koaid          ... frameno
         object         ...  int32 
 ---------------------- ... -------
 HI.20040824.08204.fits ...    2300
 HI.20040824.08544.fits ...    2301
 HI.20040824.08591.fits ...    2302)

In [86]:
table = results.to_table()
table.write ('./maxrec_3.tbl', format='ascii.ipac', overwrite=True)

### PASS 

### Test 9: set max records =  does not exist


In [91]:
sql="select koaid, filehand, frameno from koa_hires where koaid like '%20040824%'"

try:
    results = koa.run_sync(sql, maxrec=DOESNOTEXIST)
    print, results
except Exception as e:
        print, {str(e)}
    


### PASS

###  Test 10 - where_1  (date range)
#### select%20koaid,%20filehand%20from%20koa.koa_hires_v7%20where%20(utdatetime%20%3E=%20to_date(%272009-01-01%2000:00:00%27,%20%27yyyy-mm-dd%20HH24:MI:SS%27)%20and%20utdatetime%20%3C=%20to_date(%272009-01-05%2023:59:59%27,%20%27yyyy-mm-dd%20HH24:MI:SS%27))%20&format=ipac

In [92]:
sql="select koaid, filehand from koa_hires where (utdatetime >= to_date('2009-01-01 00:00:00', 'yyyy-mm-dd HH24:MI:SS') and utdatetime <= to_date('2009-01-05 23:59:59', 'yyyy-mm-dd HH24:MI:SS'))"
results  = koa.run_sync(sql)
print, results

(<function print>, <Table length=61>
         koaid                                filehand                      
         object                                object                       
 ---------------------- ----------------------------------------------------
 HI.20090101.09133.fits /koadata4/HIRES/20090101/lev0/HI.20090101.09133.fits
 HI.20090101.09658.fits /koadata4/HIRES/20090101/lev0/HI.20090101.09658.fits
 HI.20090101.09784.fits /koadata4/HIRES/20090101/lev0/HI.20090101.09784.fits
 HI.20090104.02004.fits /koadata4/HIRES/20090104/lev0/HI.20090104.02004.fits
 HI.20090104.02222.fits /koadata4/HIRES/20090104/lev0/HI.20090104.02222.fits
 HI.20090104.02477.fits /koadata4/HIRES/20090104/lev0/HI.20090104.02477.fits
 HI.20090105.04559.fits /koadata4/HIRES/20090105/lev0/HI.20090105.04559.fits
 HI.20090105.05148.fits /koadata4/HIRES/20090105/lev0/HI.20090105.05148.fits
 HI.20090105.06188.fits /koadata4/HIRES/20090105/lev0/HI.20090105.06188.fits
 HI.20090105.08445.fits /koadata4/HIRES

In [93]:

  
## write to output file
table_daterange=results.to_table()
table.write ('./table_daterange.tbl',format='votable',overwrite=True)

### Pass

## Test 11: select program ID

In [94]:
sql="select koaid, filehand, progid, iodin, imagetyp from koa_hires where (progid = 'C307') "  
results  = koa.run_sync(sql)
print, results

(<function print>, <Table length=538>
         koaid          ... imagetyp
         object         ...  object 
 ---------------------- ... --------
 HI.20180331.02525.fits ...    focus
 HI.20180331.03342.fits ...    focus
 HI.20180331.03598.fits ...    focus
 HI.20180331.03786.fits ...    focus
 HI.20180331.03943.fits ...    focus
 HI.20180331.04060.fits ...    focus
 HI.20180331.04218.fits ...    focus
 HI.20180331.04338.fits ...    focus
 HI.20180331.04502.fits ...  arclamp
 HI.20180331.04578.fits ...  arclamp
                    ... ...      ...
 HI.20180801.45058.fits ...   object
 HI.20180801.45222.fits ...   object
 HI.20180801.45356.fits ...   object
 HI.20180801.45644.fits ...   object
 HI.20180801.45867.fits ...   object
 HI.20180801.45945.fits ...   object
 HI.20180801.46157.fits ... flatlamp
 HI.20180801.46246.fits ... flatlamp
 HI.20180801.46399.fits ...  arclamp
 HI.20180801.46463.fits ...  arclamp)

In [95]:
## write to output file
table_progID_C307=results.to_table()
table.write ('./table_priogID_C307.tbl',format='votable',overwrite=True)

### PASS


## Test 12: Spatial_1 - circle centered on RA = 262.0 Dec = 17.01   Radius 1.0   units in decimal degrees

In [96]:
sql=("select koaid, filehand, ra2000, dec2000 from koa_hires where contains(point('icrs', ra2000, dec2000), circle('icrs',262.0,17.0,1.0)) = 1")

In [97]:
results  = koa.run_sync(sql)
print, results

(<function print>, <Table length=42>
         koaid          ...  dec2000 
         object         ...  float64 
 ---------------------- ... ---------
 HI.19980718.34975.fits ... 16.827111
 HI.19980912.26279.fits ... 16.824889
 HI.19980913.27523.fits ... 16.822694
 HI.19980914.23928.fits ... 16.821861
 HI.19980914.24204.fits ... 16.821861
 HI.19980914.29108.fits ... 16.822222
 HI.19980915.26653.fits ...   16.8225
 HI.19980916.25837.fits ... 16.823694
 HI.19980918.24196.fits ... 16.825194
 HI.19990421.49245.fits ...   17.5081
                    ... ...       ...
 HI.20040824.18925.fits ... 17.509944
 HI.20040825.18800.fits ... 17.510194
 HI.20041002.18948.fits ... 17.509944
 HI.20041002.19291.fits ... 17.509944
 HI.20041002.19935.fits ... 17.509944
 HI.20041002.20579.fits ... 17.509944
 HI.20050301.49856.fits ... 16.825028
 HI.20160126.56113.fits ...   16.8245
 HI.20160714.38961.fits ...  16.82319
 HI.20180702.37596.fits ...  16.82131)

### PASS

### Test 13 - Spatial_2 box with 262.0 Dec = 17.01    and 1 degree 

In [98]:
sql=("select koaid, filehand, ra2000, dec2000 from koa_hires where contains(point('icrs', ra2000, dec2000), box('icrs',262.0,17.0,2.0,2.0)) = 1")

In [99]:
results = koa.run_sync(sql)
print, results

(<function print>, <Table length=42>
         koaid          ...  dec2000 
         object         ...  float64 
 ---------------------- ... ---------
 HI.19980718.34975.fits ... 16.827111
 HI.19980912.26279.fits ... 16.824889
 HI.19980913.27523.fits ... 16.822694
 HI.19980914.23928.fits ... 16.821861
 HI.19980914.24204.fits ... 16.821861
 HI.19980914.29108.fits ... 16.822222
 HI.19980915.26653.fits ...   16.8225
 HI.19980916.25837.fits ... 16.823694
 HI.19980918.24196.fits ... 16.825194
 HI.19990421.49245.fits ...   17.5081
                    ... ...       ...
 HI.20040824.18925.fits ... 17.509944
 HI.20040825.18800.fits ... 17.510194
 HI.20041002.18948.fits ... 17.509944
 HI.20041002.19291.fits ... 17.509944
 HI.20041002.19935.fits ... 17.509944
 HI.20041002.20579.fits ... 17.509944
 HI.20050301.49856.fits ... 16.825028
 HI.20160126.56113.fits ...   16.8245
 HI.20160714.38961.fits ...  16.82319
 HI.20180702.37596.fits ...  16.82131)

In [100]:
## write to output file
table_spatial_1=results.to_table()
table.write ('./table_spatial_1.tbl',format='votable',overwrite=True)

#### FAIL - no records returned; known bug

### Test 14 - spatial_3, polygon

In [101]:
sql=("select koaid, filehand, ra2000, dec2000 from koa.koa_hires_v7 where contains(point('icrs', ra2000, dec2000), polygon('icrs',209.80225,53.34894,209.80225,55.34894,211.80225,54.34894)) = 1")

In [102]:
results = koa.run_sync(sql)
print, results

(<function print>, <Table length=28>
         koaid          ...  dec2000 
         object         ...  float64 
 ---------------------- ... ---------
 HI.20040114.51275.fits ... 54.353806
 HI.20040114.53552.fits ... 54.353806
 HI.20040114.55829.fits ... 54.353806
 HI.20040114.58182.fits ... 54.353833
 HI.20060420.33559.fits ... 54.352917
 HI.20060420.35412.fits ... 54.352917
 HI.20060420.37265.fits ... 54.352917
 HI.20060420.39117.fits ... 54.352917
 HI.20060420.41232.fits ... 54.317028
 HI.20060420.43085.fits ... 54.317028
                    ... ...       ...
 HI.20110905.19585.fits ...  54.27675
 HI.20110909.17581.fits ...  54.27289
 HI.20110909.18064.fits ...  54.27289
 HI.20110909.18415.fits ...  54.27289
 HI.20110909.19067.fits ...  54.27289
 HI.20110909.19719.fits ...  54.27289
 HI.20110920.17641.fits ...  54.27347
 HI.20110923.18047.fits ...  54.27414
 HI.20110923.18126.fits ...  54.27414
 HI.20110923.18478.fits ...  54.27414)

In [103]:
## write to output file
table_spatial_P=results.to_table()
table.write ('./table_spatial_P.tbl',format='votable',overwrite=True)

In [104]:
print, results

(<function print>, <Table length=28>
         koaid          ...  dec2000 
         object         ...  float64 
 ---------------------- ... ---------
 HI.20040114.51275.fits ... 54.353806
 HI.20040114.53552.fits ... 54.353806
 HI.20040114.55829.fits ... 54.353806
 HI.20040114.58182.fits ... 54.353833
 HI.20060420.33559.fits ... 54.352917
 HI.20060420.35412.fits ... 54.352917
 HI.20060420.37265.fits ... 54.352917
 HI.20060420.39117.fits ... 54.352917
 HI.20060420.41232.fits ... 54.317028
 HI.20060420.43085.fits ... 54.317028
                    ... ...       ...
 HI.20110905.19585.fits ...  54.27675
 HI.20110909.17581.fits ...  54.27289
 HI.20110909.18064.fits ...  54.27289
 HI.20110909.18415.fits ...  54.27289
 HI.20110909.19067.fits ...  54.27289
 HI.20110909.19719.fits ...  54.27289
 HI.20110920.17641.fits ...  54.27347
 HI.20110923.18047.fits ...  54.27414
 HI.20110923.18126.fits ...  54.27414
 HI.20110923.18478.fits ...  54.27414)

#### Pass


### Test 15 - count between 20090101 and 20090105, do count, ipac table format 

In [105]:
sql=("select count(*) from koa_hires where (utdatetime >= to_date('2009-01-01 00:00:00', 'yyyy-mm-dd HH24:MI:SS') and utdatetime <= to_date('2009-01-05 23:59:59', 'yyyy-mm-dd HH24:MI:SS'))")

In [106]:
results = koa.run_sync(sql)

In [107]:
print,results


(<function print>, <Table length=1>
 count(*)
  int32  
 --------
       61)

#### Re-run and call count(*) as total


In [108]:
sql=("select count(*) as total from koa_hires where (utdatetime >= to_date('2009-01-01 00:00:00', 'yyyy-mm-dd HH24:MI:SS') and utdatetime <= to_date('2009-01-05 23:59:59', 'yyyy-mm-dd HH24:MI:SS'))")

In [109]:
results = koa.run_sync(sql)

In [110]:
print, results

(<function print>, <Table length=1>
 total
 int32
 -----
    61)

#### Pass

### Test 16 - full table count

In [111]:
sql="select count(*) as total from koa_hires"

In [112]:
results = koa.run_sync(sql)

In [113]:
print,results

(<function print>, <Table length=1>
 total 
 int32 
 ------
 408339)

#### PASS

### Test  17 - count on a keyword that does not exist

In [116]:
sql="select count(DOESNOTEXIST) from koa_hires where (utdatetime >= to_date('2009-01-01 00:00:00', 'yyyy-mm-dd HH24:MI:SS') and utdatetime <= to_date('2009-01-05 23:59:59', 'yyyy-mm-dd HH24:MI:SS'))"

In [117]:
try:
    results = koa.run_sync(sql)
    table   = results.to_table()
    table.write('./select_countnokeyword.tbl',format='ascii.ipac',overwrite=True)
    
except Exception as e:
    with open ('./select_countnokeyword.tbl', 'w+') as fp:
      fp.write (f'Error: {str(e):s}')
      fp.flush()

### Pass - reports error message  


In [119]:
sql=("select airmass from koa_hires")

results = koa.run_sync(sql)

print(results)

<Table length=408339>
airmass
float64
-------
   1.62
   1.77
   1.56
   1.66
   1.62
   1.71
   1.01
   1.43
   1.64
   5.14
    ...
   1.99
   1.99
   1.99
   1.99
   1.99
   1.99
   1.99
   1.99
   1.99
   1.99


In [120]:
sql=("select echangl  from koa_hires where contains(point('icrs', ra2000, dec2000), circle('icrs',262.0,17.0,1.0)) = 1")

results = koa.run_sync(sql)

print(results)

<Table length=42>
  echangl  
  float64  
-----------
 0.00088795
-0.00202961
-0.00202961
-0.00202961
-0.00202961
-0.00202961
-0.00046512
 -8.457e-05
 -4.228e-05
        0.0
        ...
 0.00291756
 0.00291756
-0.00291756
-0.00291756
-0.00291756
-0.00291756
-0.00291756
-0.01843559
-0.01957724
-0.00270614
